include/dol/bases/d_mj2d_data.hpp

```diff
@@ -0,0 +1,328 @@
+#pragma once
+#include <types.h>
+#include <constants/game_constants.h>
+/// @file
+
+#define SAVE_VERSION 14 ///< The current save version.
+#define SAVE_SUB_VERSION 0 ///< The current save subversion.
+
+class dSaveMng_c;
+
+/// @brief Save data header holder.
+/// @details The data stored here is temporary, as it will be discarded unless the game is saved.
+/// Size must be 32-byte aligned.
+class dMj2dHeader_c {
+    public:
+        dMj2dHeader_c(); ///< Constructs the holder.
+
+        void initialize(); ///< Initializes the header data.
+
+        /// @brief Gets the Free Mode play count for the given world/level. See ::mFreeModePlayCount.
+        /// @note Unofficial name.
+        u16 getFreeModePlayCount(int world, int level) const;
+
+        /// @brief Sets the Free Mode play count for the given world/level. See ::mFreeModePlayCount.
+        /// @note Unofficial name.
+        void setFreeModePlayCount(int world, int level, u16 count);
+
+        /// @brief Gets the Coin Battle play count for the given world/level. See ::mCoinBattlePlayCount.
+        /// @note Unofficial name.
+        u16 getCoinBattlePlayCount(int world, int level) const;
+
+        /// @brief Sets the Coin Battle play count for the given world/level. See ::mCoinBattlePlayCount.
+        /// @note Unofficial name.
+        void setCoinBattlePlayCount(int world, int level, u16 count);
+
+        /// @brief Unlocks the given world in extra modes.
+        /// @note Unofficial name.
+        void setMultiModeCompletion(int world);
+
+        /// @brief Checks if the given world is unlocked in extra modes.
+        /// @note Unofficial name.
+        bool getMultiModeCompletion(int world) const;
+
+private:
+        /// @brief Checks that the save data version matches the current one and clears the slot if not.
+        void versionUpdate();
+        void setSelectFileNo(s8 file); ///< Sets the last used save data slot.
+
+        char mMagic[4]; ///< The savegame magic.
+        u8 mVersion[2]; ///< The save data version and subversion. See ::SAVE_VERSION and ::SAVE_SUB_VERSION.
+        u8 mLastSelectedFile; ///< The last selected save data slot.
+        // [1 byte padding]
+
+        /// @brief The play count of each level in Free Mode.
+        /// @details [Value is increased by 0x100 for every playthrough until 10000].
+        u16 mFreeModePlayCount[WORLD_COUNT][STAGE_COUNT];
+
+        /// @brief The play count of each level in Coin Battle.
+        /// @details [Value is increased by 0x100 for every playthrough until 10000].
+        u16 mCoinBattlePlayCount[WORLD_COUNT][STAGE_COUNT];
+
+        /// @brief The worlds unlocked in Extra Modes.
+        /// @details Playing a world in extra modes requires said world to be unlocked in at least
+        /// one save slot. [Value is a bitfield].
```

Comment:
```suggestion
        /// one save slot. Value is a bitfield.
```
I feel like this is a reasonable thing to include in "normal" documentation too